src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs

```diff
@@ -6117,16 +6118,23 @@ private void LoadSearchValues(ReadOnlySpan<char> chars)
             // Logically do _searchValues[index], but avoid the bounds check on accessing the array,
             // and cast to the known derived sealed type to enable devirtualization.
 
-            // DerivedSearchValues d = Unsafe.Add(ref MemoryMarshal.GetArrayDataReference(this._searchValues), index);
+            // DerivedSearchValues d = Unsafe.As<DerivedSearchValues>(Unsafe.Add(ref MemoryMarshal.GetArrayDataReference(this._searchValues), index));
             // ... = d;
             Ldthisfld(s_searchValuesArrayField);
             Call(s_memoryMarshalGetArrayDataReferenceSearchValues);
             Ldc(index * IntPtr.Size);
             Add();
             _ilg!.Emit(OpCodes.Ldind_Ref);
-            LocalBuilder ioavLocal = _ilg!.DeclareLocal(list[index].GetType());
-            Stloc(ioavLocal);
-            Ldloc(ioavLocal);
+            Call(MakeUnsafeAs(list[index].GetType())); // provide JIT with details necessary to devirtualize calls on this instance
+
+            [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2060:MakeGenericMethod", Justification =
+                "Only uses Unsafe.As<T> with reference types <T>, which it also uses in this method to ensure it's rooted. Unsafe.As is an intrinsic with a nop " +
+                "implementation supplied at codegen time. And this is also part of the regex compiler that's only used when dynamic code compilation is supported.")]
```

Comment:
@eerhardt, does this look ok?

Reply:
> But maybe because the MethodInfo is a static variable, the trimmer doesn't recognize the pattern?

This is correct - trimmer can't see through static fields right now, so it treats this as calling `MakeGenericMethod` on a completely unknown `MethodInfo`, and thus it has to assume the worst, which is that it might have some annotations.